src/core/metrics.cc

```diff
@@ -156,205 +156,215 @@ Metrics::EnableGPUMetrics()
   }
 
   if (std::getenv("TRITON_SERVER_CPU_ONLY") == nullptr) {
-    singleton->InitializeNvmlMetrics();
+    singleton->InitializeDcgmMetrics();
   }
 
   singleton->gpu_metrics_enabled_ = true;
 }
 
 bool
-Metrics::InitializeNvmlMetrics()
+Metrics::InitializeDcgmMetrics()
 {
 #ifndef TRITON_ENABLE_METRICS_GPU
   return false;
 #else
-  nvmlReturn_t nvmlerr = nvmlInit();
-  if (nvmlerr != NVML_SUCCESS) {
-    LOG_WARNING << "failed to initialize, GPU metrics will not be available: "
-                << nvmlErrorString(nvmlerr);
+  dcgmReturn_t dcgmerr = dcgmInit();
+  if (dcgmerr != DCGM_ST_OK) {
+    LOG_WARNING
+        << "failed to initialize DCGM, GPU metrics will not be available: "
+        << errorString(dcgmerr);
     return false;
   }
 
-  int dcnt;
-  cudaError_t cudaerr = cudaGetDeviceCount(&dcnt);
-  if (cudaerr != cudaSuccess) {
-    LOG_WARNING
-        << "failed to get device count, GPU metrics will not be available: "
-        << cudaGetErrorString(cudaerr);
+  dcgmerr = dcgmStartEmbedded(DCGM_OPERATION_MODE_MANUAL, &dcgm_handle_);
+  if (dcgmerr != DCGM_ST_OK) {
+    LOG_WARNING << "Error: dcgmStartEmbedded returned \""
+                << errorString(dcgmerr) << "(" << dcgmerr << ")\"";
     return false;
   }
 
-  // Create NVML metrics for each GPU
-  for (int didx = 0; didx < dcnt; ++didx) {
-    // Get handle for the GPU
-    cudaDeviceProp gpu_properties;
-    cudaerr = cudaGetDeviceProperties(&gpu_properties, didx);
-    if (cudaerr != cudaSuccess) {
-      LOG_WARNING << "failed to get device properties for device " << didx
-                  << ", GPU metrics will not be available for this device: "
-                  << cudaGetErrorString(cudaerr);
-      continue;
-    }
+  dcgmerr = dcgmUpdateAllFields(dcgm_handle_, 1);
+  if (dcgmerr != DCGM_ST_OK) {
+    LOG_WARNING << "DCGM failed to update all fields, GPU metrics will not "
+                   "be available: "
+                << errorString(dcgmerr);
+    return false;
+  }
 
-    char pcibusid_str[64];
-    cudaerr =
-        cudaDeviceGetPCIBusId(pcibusid_str, sizeof(pcibusid_str) - 1, didx);
-    if (cudaerr != cudaSuccess) {
-      LOG_WARNING << "failed to get Bus ID for device " << didx
-                  << ", GPU metrics will not be available for this device: "
-                  << cudaGetErrorString(cudaerr);
-      continue;
-    }
+  unsigned int all_gpu_ids[DCGM_MAX_NUM_DEVICES];
+  int gpu_count;
+  dcgmerr = dcgmGetAllDevices(dcgm_handle_, all_gpu_ids, &gpu_count);
```

Comment:
Does DCGM only return the cuda visible devices so the filtering is not needed?

Reply:
Yes, DCGM will report a list of all visible devices but it is possible dcgmGetDeviceAttributes is unable to get the properties for some of those devices

Reply:
To be clear, by visible devices I mean only those visisble by CUDA due to CUDA_VISIBLE_DEVICES (and NVIDIA_VISIBLE_DEVICES in containers)